pinot-plugins/pinot-file-system/pinot-s3/src/main/java/org/apache/pinot/plugin/filesystem/S3PinotFS.java

```diff
@@ -0,0 +1,485 @@
+
+                if (prefix.equals(DELIMITER)) {
+                    ListObjectsV2Request listObjectsV2Request = listObjectsV2RequestBuilder.build();
+                    listObjectsV2Response = s3Client.listObjectsV2(listObjectsV2Request);
+                } else {
+                    ListObjectsV2Request listObjectsV2Request = listObjectsV2RequestBuilder.prefix(prefix).build();
+                    listObjectsV2Response = s3Client.listObjectsV2(listObjectsV2Request);
+                }
+                boolean deleteSucceeded = true;
+                for (S3Object s3Object : listObjectsV2Response.contents()) {
+                    DeleteObjectRequest deleteObjectRequest = DeleteObjectRequest.builder()
+                            .bucket(segmentUri.getHost())
+                            .key(s3Object.key())
+                            .build();
+
+                    DeleteObjectResponse deleteObjectResponse = s3Client.deleteObject(deleteObjectRequest);
+
+                    deleteSucceeded &= deleteObjectResponse.sdkHttpResponse().isSuccessful();
+                }
+                return deleteSucceeded;
+            } else {
+                String prefix = sanitizePath(segmentUri.getPath());
+                DeleteObjectRequest deleteObjectRequest = DeleteObjectRequest.builder()
+                        .bucket(segmentUri.getHost())
+                        .key(prefix)
+                        .build();
+
+                DeleteObjectResponse deleteObjectResponse = s3Client.deleteObject(deleteObjectRequest);
+
+                return deleteObjectResponse.sdkHttpResponse().isSuccessful();
+            }
+        }catch(NoSuchKeyException e){
+            return false;
+        } catch (S3Exception e) {
+            throw e;
+        } catch(Throwable t) {
+            throw new IOException(t);
+        }
+    }
+
+    @Override
+    public boolean doMove(URI srcUri, URI dstUri) throws IOException {
+        if (copy(srcUri, dstUri)) {
+            return  delete(srcUri, true);
+        }
+        return false;
+    }
+
+    @Override
+    public boolean copy(URI srcUri, URI dstUri) throws IOException {
+        LOGGER.info("Copying uri {} to uri {}", srcUri, dstUri);
+        checkState(exists(srcUri), "Source URI '%s' does not exist", srcUri);
+        if (srcUri.equals(dstUri)) {
+            return true;
+        }
+        if (!isDirectory(srcUri)) {
+            delete(dstUri, true);
+            return copyFile(srcUri, dstUri);
+        }
+        dstUri = normalizeToDirectoryUri(dstUri);
+        ImmutableList.Builder<URI> builder = ImmutableList.builder();
+        Path srcPath = Paths.get(srcUri.getPath());
+        try {
+            boolean copySucceeded = true;
+            for (String directoryEntry : listFiles(srcUri, true)) {
+                URI src = new URI(srcUri.getScheme(), srcUri.getHost(), directoryEntry, null);
+                String relativeSrcPath = srcPath.relativize(Paths.get(directoryEntry)).toString();
```

Comment:
Could you add a test to ensure this copy directory logic is correct?
Especially the relative path is correct.

Reply:
done